File: nim/day02/solve.nim
import std/[strformat, strutils, sequtils]
import move

func part1(moves: openArray[Move]): int =
  var h_pos = 0
  var depth = 0
  for move in moves:
    case move.kind:
    of down: depth += move.down
    of forward: h_pos += move.forward
  h_pos * depth

func part2(moves: openArray[Move]): int =
  var h_pos = 0
  var depth = 0
  var aim = 0
  for move in moves:
    case move.kind:
    of down: aim += move.down
    of forward:
      h_pos += move.forward
      depth += move.forward * aim
  h_pos * depth

proc solve* =
  let moves = ".input/day02".readFile.strip.splitLines.map parseMove
  let part1 = moves.part1
  let part2 = moves.part2
  echo &"day02: {part1} {part2}"

File: c++/day02/solve.cpp
#include <iostream>
#include "move.cpp"
#include "../util.hpp"

namespace day02 {
  auto solve_part1(std::vector<Move>& moves) -> int32_t {
    int32_t h_pos = 0;
    int32_t depth = 0;
    for (auto move : moves) {
      switch (move.kind) {
      case Down: depth += move.down; break;
      case Forward: h_pos += move.forward; break;
      }
    }
    return h_pos * depth;
  }

  auto solve_part2(std::vector<Move>& moves) -> int32_t {
    int32_t h_pos = 0;
    int32_t depth = 0;
    int32_t aim = 0;
    for (auto move : moves) {
      switch (move.kind) {
      case Down: aim += move.down; break;
      case Forward:
        h_pos += move.forward;
        depth += move.forward * aim;
        break;
      }
    }
    return h_pos * depth;
  }

  auto solve() -> void {
    auto input = util::read_file_to_string(".input/day02");
    auto moves = Move::parse_many(input);

    auto part1 = solve_part1(moves);
    auto part2 = solve_part2(moves);

    std::cout << "day02: " << part1 << ' ' << part2 << '\n';
  }
}
